src/backend/distributed/utils/ruleutils_12.c

```diff
@@ -0,0 +1,7913 @@
+/*-------------------------------------------------------------------------
```

Comment:
I'm done reviewing `ruleutils_12.c`, the way you created this file makes sense and I've double checked if anything missed. As I noted in the other issue, adding some tests with `MATERIALIZED` and `NOT MATERIALIZED` would be nice.